src/Microsoft.ServiceHub.Framework/Container/GlobalBrokeredServiceContainer.cs

```diff
@@ -269,6 +269,23 @@ public IServiceBroker GetSecureServiceBroker(ServiceActivationOptions options)
 	/// </devremarks>
 	public ValueTask<TraceSource?> GetTraceSourceForBrokeredServiceAsync(IServiceBroker serviceBroker, ServiceMoniker serviceMoniker, ServiceActivationOptions options, bool clientRole, CancellationToken cancellationToken) => this.GetTraceSourceForConnectionAsync(serviceBroker, serviceMoniker, options, clientRole, cancellationToken);
 
+	/// <summary>
+	/// Applies typical transformations on a descriptor fro brokered service clients and services.
+	/// </summary>
+	/// <param name="descriptor">The stock descriptor used for this service.</param>
+	/// <param name="serviceBroker">A service broker that may be used to acquire other, related services as necessary to mutate the <paramref name="descriptor"/>.</param>
+	/// <param name="serviceActivationOptions">The activation options for the service.</param>
+	/// <param name="clientRole">A value indicating whether the <paramref name="descriptor"/> is about to be used to activate a client proxy or client connection; use <see langword="false" /> when activating the service itself.</param>
+	/// <param name="cancellationToken">A cancellation token.</param>
+	/// <returns>The modified descriptor.</returns>
+	internal async ValueTask<ServiceRpcDescriptor> ApplyDescriptorSettingsAsync(ServiceRpcDescriptor descriptor, IServiceBroker serviceBroker, ServiceActivationOptions serviceActivationOptions, bool clientRole, CancellationToken cancellationToken)
```

Comment:
Can this be made protected virtual so that implementations like those in VS can also do their own thing with the descriptor? Otherwise those implementations don't get any chance to observe/manipulate the descriptor.

Reply:
I was thinking of you when I added this, and almost did it. But then I had to add 3 more parameters to it. Then I had to make it async, and I realized... there's no rhyme or reason at all to the information I'm providing to this method except implementation detail. We could easily find we need another parameter next week added to it. So I'm not comfortable yet with exposing the method publicly from the assembly (even as `protected`). I especially feel that the async return type is gross and would like to get rid of it. But at the moment our TraceHub implementation requires it. If we can fix that, I'll change this method to be a synchronous method and probably remove IServiceBroker as a parameter. 

Besides, it's still not clear to me how you would mutate the descriptor to serve your needs. They can only be changed in a very few ways, which the container currently takes care of.